### ISSUE TITLE
mod_push wrongly tries to push chat status messages

### ISSUE THREAD

**Issue by rspq** [esl/MongooseIM#1406]
MongooseIM version: 2.1
Installed from: Docker
Erlang/OTP version: 18

I am trying to get push notifications working and observed the below crash in the log. I believe it is caused by mod_push assuming that all messages have body elements, a false assumption because chat status messages do not have a body.

```
[mongooseim-1] 2017-07-18 17:00:34.625 [error] <0.5563.5>@ejabberd_hooks:run_fold1:263 {function_clause,[{exml_query,cdata,[undefined],[{file,"/home/travis/build/esl/MongooseIM/_build/default/lib/exml/src/exml_query.erl"},{line,112}]},{mod_push,push_notification_iq,5,
[{file,"/home/travis/build/esl/MongooseIM/_build/prod/lib/ejabberd/src/mod_push.erl"},{line,255}]},{mod_push,'-publish_message/3-fun-3-',6,[{file,"/home/travis/build/esl/MongooseIM/_build/prod/lib/ejabberd/src/mod_push.erl"},{line,178}]},{lists,foreach,2,[{file,"lists.erl"},{line,1337}]},{mod_push,publish_message,3,[{file,"/home/travis/build/esl/MongooseIM/_build/prod/lib/ejabberd/src/mod_push.erl"},{line,176}]},{mod_push,filter_packet,1,[{file,"/home/travis/build/esl/MongooseIM/_build/prod/lib/ejabberd/src/mod_push.erl"},{line,124}]},{safely,apply,3,[{file,"/home/travis/build/esl/MongooseIM/_build/prod/lib/ejabberd/src/safely.erl"},{line,19}]},{ejabberd_hooks,hook_apply_function,5,[{file,"/home/travis/build/esl/MongooseIM/_build/prod/lib/ejabberd/src/ejabberd_hooks.erl"},{line,277}]}]} 
 mongooseim-1] running hook: {filter_local_packet,[]}
```

EDIT: Something about the Erlang stack trace is upsetting the code element above. You'll need to look at the source of this post to see the whole thing.

**Comment by hartmut-pq** [comment]
Hi, we (colleague of @rspq) can't get push working at all. 
Followed docs precisely, MongoosePush is up and running.
We can't get any error messages that would help to indicate any issues other than the one mentioned above.

*Example accessing MongoosePush from our MongooseIM instance:*
```
root@mongooseim-0:/# curl -X GET http://mongoosepush
{"details":"This is not the endpoint you are looking for."}root@mongooseim-0:/#
root@mongooseim-0:/# curl -X POST http://mongoosepush/v1/notification/f534534543
{"details":"Parsing Param Error: service"}root@mongooseim-0:/#
```

Relevant part of our ejabberd.cfg:
```
<...>

{http_connections, [{mongoose_push_http,
    [{server, "http://mongoosepush"}]
}]}.

<...>

  {mod_pubsub, [
     {plugins, [<<"push">>]}
  ]},
  {mod_push, [
     {backend, mnesia},
     {wpool, [{workers, 200}]},
     {plugin_module, mod_push_plugin_default}
  ]},
  {mod_push_service_mongoosepush, [
      {pool_name, mongoose_push_http},
      {api_version, "v1"}
  ]},

<...>
```

**Comment by hartmut-pq** [comment]
Just found https://github.com/esl/MongooseIM/pull/1408
Could this be related @fenek @rslota ?

**Comment by rslota** [comment]
Hi @hartmut-pq,

There are two different issues here. The bug reported by @rspq is indeed just a bug that generates error logs that shouldn't be logged as errors, just ignored.

The issue you're having on the other hand is due to "typo-grade" bug in the documentation. I have already opened PR - #1408 that fixes the issue. Please change your config accordingly.

**Comment by hartmut-pq** [comment]
Thx, we'll give it a try...

Getting
```
 mongooseim-1] running hook: {filter_local_packet,[]}
[mongooseim-1] 2017-07-25 08:54:22.147 [error] <0.30118.3>@ejabberd_hooks:run_fold1:263 {{case_clause,undefined},[{mod_push_plugin_default,sender_id,2,[{file,"/home/travis/build/esl/MongooseIM/_build/prod/lib/ejabberd/src/mod_push_plugin_default.erl"},{line,51}]},{mod_push,push_notification_iq,5,[{file,"/home/travis/build/esl/MongooseIM/_build/prod/lib/ejabberd/src/mod_push.erl"},{line,254}]},{mod_push,'-publish_message/3-fun-3-',6,[{file,"/home/travis/build/esl/MongooseIM/_build/prod/lib/ejabberd/src/mod_push.erl"},{line,178}]},{lists,foreach,2,[{file,"lists.erl"},{line,1337}]},{mod_push,publish_message,3,[{file,"/home/travis/build/esl/MongooseIM/_build/prod/lib/ejabberd/src/mod_push.erl"},{line,176}]},{mod_push,filter_packet,1,[{file,"/home/travis/build/esl/MongooseIM/_build/prod/lib/ejabberd/src/mod_push.erl"},{line,124}]},{safely,apply,3,[{file,"/home/travis/build/esl/MongooseIM/_build/prod/lib/ejabberd/src/safely.erl"},{line,19}]},{ejabberd_hooks,hook_apply_function,5,[{file,"/home/travis/build/esl/MongooseIM/_build/prod/lib/ejabberd/src/ejabberd_hooks.erl"},{line,277}]}]}
```

and can't really make much sense of it. Sth wrong with the config?

**Comment by rslota** [comment]
This error means that somehow push notification was generated based on message that don't have `type` attribute (which should b either `chat` or `groupchat`).

**Comment by hartmut-pq** [comment]
So is this one of the server-side errors that can be ignored? Or sth wrong with the client?

**Comment by rslota** [comment]
This is something slightly different, a bug on the server, because `type` attribute is optional. Although if `type` is missing, the server has to assume it's a `normal` message which is not supported by Push Notifications. It has to be either a `chat` or `groupchat` message in which case `type` MUST be specified and therefore crash you are seeing cannot happen.

**Comment by hartmut-pq** [comment]
Thx a lot @rslota for the support. We've got some push messages sent and received now - so seems to have been resolved with the `v2` config change.

**Comment by fenek** [comment]
Hi @hartmut-pq

Regarding the error you were seeing: Was it still with our official Docker image? Or a MongooseIM instance you've customised (i.e. modified the source code)?

**Comment by hartmut-pq** [comment]
@fenek ~official `mongooseim/mongooseim:PR-1391`

**Comment by rslota** [comment]
@fenek @hartmut-pq ,

There is no need for investigating this issue. I already confirmed it a while ago and it's gonna by fixed by #1414 .

**Comment by shanbhardwaj** [comment]
+1 guys this is really killing us. We are not able to reliably get our push infrastructure up and running.

**Comment by fenek** [comment]
Hi @shanbhardwaj 

I'll do my best to merge the PR mentioned by @rslota soon but unfortunately I can't promise when exactly...

**Comment by izigibran** [comment]
+1 I would love this fix also

**Comment by deadjdona** [comment]
#1414 is merged now.

**Comment by fenek** [comment]
Fixed by #1726
